language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/Function.hs

```diff
@@ -92,3 +95,132 @@ getBuiltinFix = do
         $ [ Apply () (unrollFunAB unroll2) $ Var () s
           , Var () x
           ]
+
+-- | A type that looks a bit transformation.
+--
+-- > trans F G Q : F Q -> G Q
+trans :: Type TyName () -> Type TyName () -> Type TyName () -> Quote (Type TyName ())
+trans f g q = pure $ TyFun () (TyApp () f q) (TyApp () g q)
+
+-- | A type that looks like a natural transformation.
+--
+-- > natTrans F G : forall Q :: * . F Q -> G Q
+natTrans :: Type TyName () -> Type TyName () -> Quote (Type TyName ())
+natTrans f g = freshTyName () "Q" >>= \q -> TyForall () q (Type ()) <$> trans f g (TyVar () q)
+
+-- | A type that looks like a natural transformation to Id.
+--
+-- > natTransId F : forall Q :: * . F Q -> Q
+natTransId :: Type TyName () -> Quote (Type TyName ())
+natTransId f = do
+    q <- freshTyName () "Q"
+    pure $ TyForall () q (Type ()) (TyFun () (TyApp () f (TyVar () q)) (TyVar () q))
+
+-- | The 'fixBy' combinator.
+--
+-- fixBy :
+--     forall F :: (* -> *) .
+--     ((forall Q :: * . F Q -> Q) -> (forall Q :: * . F Q -> Q)) ->
+--     ((forall Q :: * . F Q -> F Q) -> (forall Q :: * . F Q -> Q))
```

Comment:
Maybe we should write `fixBy : forall (F :: * -> *). ((F ~> Id) -> F ~> Id) -> (F ~> F) -> F ~> Id`? Kinda shorter...
Also the `forall F :: (* -> *)` syntax looks somewhat nonstandard to me, I'd much prefer `forall (F :: * -> *)`.

Reply:
I like it!